include/nbl/builtin/hlsl/subgroup/fft.hlsl

```diff
@@ -12,49 +12,94 @@ namespace hlsl
 {
-    const vector <Scalar, 4> loHiPacked = {lo.real(), lo.imag(), hi.real(), hi.imag()};
-    vector <Scalar, 4> shuffledLoHiPacked = glsl::subgroupShuffleXor< vector <Scalar, 4> > (loHiPacked, stride);
-    lo.real(shuffledLoHiPacked.x);
-    lo.imag(shuffledLoHiPacked.y);
-    hi.real(shuffledLoHiPacked.z);
-    hi.imag(shuffledLoHiPacked.w);
-    // Get twiddle with k = subgroupID mod stride, N = 2 * stride
-    fft::DIX<Scalar, inverse>::radix2(fft::twiddle<Scalar, inverse>(glsl::gl_SubgroupID() & (stride - 1), stride << 1), lo, hi);   
-}
+    static void __call(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi);
+};
+
+// ---------------------------------------- Radix 2 forward transform - DIF -------------------------------------------------------
 
-// When doing a 1D FFT of size N, we call N/2 threads to do the work, where each thread of index 0 <= k < N/2 is in charge of computing two elements
-// in the output array, those of index k and k + N/2. The first is "lo" and the latter is "hi"
-template<typename Scalar, uint32_t SubgroupSize, bool inverse>
-void FFT(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi) 
+template<typename Scalar, class device_capabilities>
+struct FFT<2, false, Scalar, device_capabilities>
 {
-    const uint32_t doubleSubgroupSize = SubgroupSize << 1;
+    static void FFT_loop(uint32_t stride, NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi)
+    {
+        const bool topHalf = (glsl::gl_SubgroupInvocationID() & stride) != 0;
+        const vector <Scalar, 2> toTrade = topHalf ? vector <Scalar, 2>(lo.real(), lo.imag()) : vector <Scalar, 2>(hi.real(), hi.imag());
+        const vector <Scalar, 2> exchanged = glsl::subgroupShuffleXor< vector <Scalar, 2> > (toTrade, stride);
+        if (topHalf)
+        {
+            lo.real(exchanged.x);
+            lo.imag(exchanged.y);
+        }
+        else
+        {
+            hi.real(exchanged.x);
+            hi.imag(exchanged.y);
+        }
+        // Get twiddle with k = subgroupInvocation mod stride, N = 2 * stride
+        fft::DIF<Scalar>::radix2(fft::twiddle<false, Scalar>(glsl::gl_SubgroupInvocationID() & (stride - 1), stride << 1), lo, hi); 
+    }
+
+    static void __call(NBL_REF_ARG(complex_t<Scalar>) lo, NBL_REF_ARG(complex_t<Scalar>) hi) 
+    {
+        const uint32_t subgroupSize = glsl::gl_SubgroupSize();  //This is N/2
+        const uint32_t doubleSubgroupSize = subgroupSize << 1;  //This is N
     
-    // special first iteration
-    if (! inverse)
-        fft::DIF<Scalar>::radix2(fft::twiddle<Scalar, false>(glsl::gl_SubgroupID(), doubleSubgroupSize), lo, hi);                                                                                   
+        // special first iteration
+        fft::DIF<Scalar>::radix2(fft::twiddle<false, Scalar>(glsl::gl_SubgroupInvocationID(), doubleSubgroupSize), lo, hi);                                                                                   
```

Comment:
why don't you define the `radix2` so it asks for `halfN` instead of `N` ? then we can skip a `<<1` at every iteration just to get your `N`